diff --git a/src/test/java/org/springframework/web/filter/Sha512ShallowEtagHeaderFilter.java b/src/test/java/org/springframework/web/filter/Sha512ShallowEtagHeaderFilter.java
index 50b2c90..8e9b1f8 100644
--- a/src/test/java/org/springframework/web/filter/Sha512ShallowEtagHeaderFilter.java
+++ b/src/test/java/org/springframework/web/filter/Sha512ShallowEtagHeaderFilter.java
@@ -5,9 +5,9 @@ import com.google.common.hash.Hashing;
 
 public class Sha512ShallowEtagHeaderFilter extends ShallowEtagHeaderFilter {
 
-	@Override
 	protected String generateETagHeaderValue(byte[] bytes) {
 		final HashCode hash = Hashing.sha512().hashBytes(bytes);
 		return "\"" + hash + "\"";
 	}
 }
+            
